pkg/cloud/openstack/openstack.go

```diff
@@ -80,3 +83,59 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
 
 	return assets, nil
 }
+
+// CloudConfigTransformer implements the cloudConfigTransformer. It uses the source ConfigMap and
+// to create a new config that adheres to the configuration for CCM
+// It returns an error if the platform is not OpenStackPlatformType.
+func CloudConfigTransformer(source string, infra *configv1.Infrastructure) (string, error) {
+	if !(infra.Status.PlatformStatus != nil &&
+		infra.Status.PlatformStatus.Type == configv1.OpenStackPlatformType) {
+		return "", fmt.Errorf("invalid platform, expected to be %s", configv1.OpenStackPlatformType)
+	}
+
+	cfg, err := ini.Load([]byte(source))
+	if err != nil {
+		return "", fmt.Errorf("failed to read the cloud.conf: %w", err)
+	}
+
+	global, _ := cfg.GetSection("Global")
+	if global != nil {
+		// Remove the legacy keys, once we ensure they're not overridden
+		for key, value := range map[string]string{
+			"secret-name":      "openstack-credentials",
+			"secret-namespace": "kube-system",
+			"kubeconfig-path":  "",
+		} {
+			if global.Key(key).String() != value {
+				return "", fmt.Errorf("'[Global] %s' is set to a non-default value", key)
+			}
+			global.DeleteKey(key)
+		}
+	} else {
+		// Section doesn't exist, ergo no validation to concern ourselves with
+		global, err = cfg.NewSection("Global")
+		if err != nil {
+			return "", fmt.Errorf("failed to modify the provided configuration: %w", err)
+		}
+	}
+
+	for key, value := range map[string]string{
+		"use-clouds":  "true",
+		"clouds-file": "/etc/openstack/secret/clouds.yaml",
+		"cloud":       "openstack",
+	} {
+		_, err = global.NewKey(key, value)
+		if err != nil {
+			return "", fmt.Errorf("failed to modify the provided configuration: %w", err)
+		}
+	}
+
```

Comment:
We should also drop everything that is in the BlockStorage section as it's now handled by the CSI driver.

On that note, it seems we're using a [static cloud.conf](https://github.com/openshift/openstack-cinder-csi-driver-operator/blob/master/assets/configmap.yaml) for Cinder CSI, and settings such as `[BlockStorage] node-volume-attach-limit` would be lost. We likely want to perform a similar transformation on the storage side.

Reply:
It seems the configuration for cloud-provider-openstack [still has a BlockStorage section](https://github.com/kubernetes/cloud-provider-openstack/blob/033052ee1ccb4986ac652facf30963a83a854d1f/pkg/openstack/openstack.go#L114-L119)? This is presumably [used by cinder CSI](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/cinder-csi-plugin/using-cinder-csi-plugin.md#block-storage). Are you sure we want to drop this?

Reply:
Done. Will work on dropping this in `cloud-provider-openstack` next.